lib/Conversion/BGVToPoly/BGVToPoly.cpp

```diff
@@ -0,0 +1,103 @@
+#include "include/Conversion/BGVToPoly/BGVToPoly.h"
+
+#include <cstddef>
+
+#include "include/Dialect/BGV/IR/BGVDialect.h"
+#include "include/Dialect/BGV/IR/BGVOps.h"
+#include "include/Dialect/BGV/IR/BGVTypes.h"
+#include "include/Dialect/Poly/IR/PolyAttributes.h"
+#include "include/Dialect/Poly/IR/PolyOps.h"
+#include "include/Dialect/Poly/IR/PolyTypes.h"
+#include "include/Dialect/Poly/IR/Polynomial.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Transforms/OneToNFuncConversions.h"
+#include "mlir/Transforms/OneToNTypeConversion.h"
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/OneToNTypeConversion.h"
+
+namespace mlir::heir::bgv {
+
+#define GEN_PASS_DEF_BGVTOPOLY
+#include "include/Conversion/BGVToPoly/BGVToPoly.h.inc"
+
+class CiphertextTypeConverter : public OneToNTypeConverter {
+ public:
+  CiphertextTypeConverter() {
+    addConversion([](Type type) { return type; });
+    addConversion([](CiphertextType type, SmallVectorImpl<Type> &results) {
+      auto ring = type.getRings().getRings()[0]; // using the first ring.
+      // TODO: figure out the right way to select ring;
```

Comment:
Hmmm. Perhaps a level int needs to be an attribute of a ciphertext, but the issue is that having it be embedded as a type attribute means that operations and functions taking BGV ciphertexts in need to validate or describe that attribute.

I think this would mean a few possibilities:
* Type attribute: if so, it should be optional (so as not to interfere with operations or subroutines). an rewrite pass would likely be able to automatically populate these, potentially adding or optimizing `modulus_switch` operations (which change level) 
* No type attribute: if so, then this lowering would require determining the level. It could do this be the following:
  1. Check the defining operation of the ciphertext. If none: assume first ring.
  2. Walk back to through parent operations: if you encounter a modulus switch, use the `to_level` attribute. (This is somewhat complicated for when the ciphertext is the result of a binary op: both parents must have belonged to the same level though, so it doesn't matter which path is taken).
  
This second alternative doesn't really seem that robust to me. Yes, we could write a pass that verifies the ciphertext leveling before and after modulus switches, and also ensures that types used in binary ops have matching levels.

Reply:
Opted to the option (1). It led to some changes in BGV_Ops, namely removed `SameOperandsAndResultType` from `BGV_ModulusSwitch`

Reply:
Right, I think I agree your choice is likely the better one here - IMO if we are specifying the ring, then we have already done some high-level analysis at that stage, and likely can infer the ciphertext level.

If not we revisit :P